src/lib/tpm.c

```diff
@@ -4008,6 +4011,111 @@ CK_RV tpm2_getmechanisms(tpm_ctx *ctx, CK_MECHANISM_TYPE *mechanism_list, CK_ULO
     return rv;
 }
 
+#ifdef HAVE_POLICY
+static TSS2_RC tpm2_policy_get_pcr(TSS2_POLICY_PCR_SELECTION *selection,
+        TPML_PCR_SELECTION *out_selection,
+        TPML_DIGEST *out_digest,
+        void *userdata)
+{
+
+    TPML_PCR_SELECTION in_pcr_selection = {0};
+    if (selection->type == TSS2_POLICY_PCR_SELECTOR_PCR_SELECTION) {
+        in_pcr_selection = selection->selections.pcr_selection;
+    } else {
+        in_pcr_selection.count = 1;
+
+        TPMS_PCR_SELECTION *pcr_bank = &in_pcr_selection.pcrSelections[0];
+        TPMS_PCR_SELECT *pcr_select = &selection->selections.pcr_select;
+
+        pcr_bank->hash = TPM2_ALG_SHA256;
+        pcr_bank->sizeofSelect = pcr_select->sizeofSelect;
+        memcpy(pcr_bank->pcrSelect, pcr_select->pcrSelect, pcr_bank->sizeofSelect);
+    }
+
+    ESYS_CONTEXT *esys_ctx = userdata;
+
+    UINT32 pcr_update_counter;
+    TPML_PCR_SELECTION *pcr_selection = NULL;
+    TPML_DIGEST *pcr_values = NULL;
+
+    TSS2_RC rc = Esys_PCR_Read(esys_ctx,
+                 ESYS_TR_NONE,
+                 ESYS_TR_NONE,
+                 ESYS_TR_NONE,
+                 &in_pcr_selection,
+                 &pcr_update_counter,
+                 &pcr_selection,
+                 &pcr_values);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOGE("Esys_PCR_Read: %s:", Tss2_RC_Decode(rc));
+        free(pcr_selection);
+        free(pcr_values);
+        return rc;
+    }
+
+    *out_selection = *pcr_selection;
+    *out_digest = *pcr_values;
+
+    free(pcr_selection);
+    free(pcr_values);
+    return TSS2_RC_SUCCESS;
+}
+
+CK_RV tpm2_execute_policy(tpm_ctx *ctx, TSS2_POLICY_CTX *policy_ctx, uint32_t handle)
+{
+
+    check_pointer(ctx);
+    check_pointer(policy_ctx);
+    check_num(handle);
+
+    TPMT_SYM_DEF symmetric = {
+        .algorithm = TPM2_ALG_AES,
+        .keyBits = { .aes = 128 },
+        .mode = { .aes = TPM2_ALG_CFB }
+    };
+
+    TSS2_POLICY_CALC_CALLBACKS calc_callbacks = {0};
+    calc_callbacks.cbpcr = &tpm2_policy_get_pcr;
+    calc_callbacks.cbpcr_userdata = ctx->esys_ctx;
+
+    TSS2_RC rc;
+
+    rc = Tss2_PolicySetCalcCallbacks(policy_ctx, &calc_callbacks);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOGE("Tss2_PolicySetCalcCallbacks: %s:", Tss2_RC_Decode(rc));
+        return CKR_GENERAL_ERROR;
+    }
+
+    /* XXX should we cache the session or running multiple policies is unlikely? */
```

Comment:
This session handle needs to be used to authorize the object for `Esys_Sign`. You're essentially still on the password model (see my other comment in the review summary). I would modify the `tpm_ctx` struct to have two session handles:
- auth_session: This is the session that will be either the HMAC session OR the policy session from execute policy.
  - For HMAC session use, it can just live the lifetime of the token.
  - For POLICY session use, it needs to be created for each command.
      - You can use the attribute continue session and use policy reset command, but thats a performance thing that can be done later.
- enc_session: This is **important**, In the current code the HMAC session handle is also the encryption handle for protecting the traffic to and from the TPM, we need this to also be created for the lifetime of the token and passed to each command as the first unused session handle, which in most commands in session2, sometimes its session3.